File: experiment-5/runs/language=java_model=claude-opus-4-8_tooling=none/rep2/src/main/java/com/brazilsoccer/mcp/model/Match.java
/*
 * ===========================================================================
 * Context: Brazilian Soccer MCP Server
 * File:    model/Match.java
 * Purpose: Immutable domain model for a single soccer match, unified across
 *          every source CSV (Brasileirão, Copa do Brasil, Libertadores, the
 *          extended BR-Football stats file and the historical 2003-2019 file).
 *          Normalized team-name keys are pre-computed at load time so query
 *          matching is fast and tolerant of naming variations (state suffixes,
 *          accents, full club names).
 * ===========================================================================
 */
package com.brazilsoccer.mcp.model;

import java.time.LocalDate;

/**
 * A single match. Goal fields may be null when a source row is missing scores.
 */
public record Match(
        String competition,
        String homeTeam,
        String awayTeam,
        String homeTeamKey,
        String awayTeamKey,
        Integer homeGoal,
        Integer awayGoal,
        Integer season,
        LocalDate date,
        String round,
        String stage,
        String source) {

    /** True when both goal values are present. */
    public boolean hasScore() {
        return homeGoal != null && awayGoal != null;
    }

    /** Total goals scored in the match, or 0 when the score is unknown. */
    public int totalGoals() {
        return hasScore() ? homeGoal + awayGoal : 0;
    }

    /**
     * Winner of the match as a normalized team key, or null for a draw or an
     * unscored match.
     */
    public String winnerKey() {
        if (!hasScore() || homeGoal.equals(awayGoal)) {
            return null;
        }
        return homeGoal > awayGoal ? homeTeamKey : awayTeamKey;
    }

    /** True when the match involves the given normalized team key (home or away). */
    public boolean involves(String teamKey) {
        return homeTeamKey.equals(teamKey) || awayTeamKey.equals(teamKey);
    }
}

File: experiments/adrianco/experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=neutral/rep1/lib/br_soccer/match.ex
defmodule BrSoccer.Match do
  @moduledoc "A single normalised match record, unified across all match CSVs."

  @enforce_keys [:competition, :source, :home, :away, :home_key, :away_key]
  defstruct [
    :competition,
    :source,
    :season,
    :date,
    :round,
    :stage,
    :home,
    :away,
    :home_key,
    :away_key,
    :home_goal,
    :away_goal,
    :home_state,
    :away_state,
    :arena,
    stats: %{}
  ]

  @type t :: %__MODULE__{}

  @doc "Whether the match has a recorded final score."
  def scored?(%__MODULE__{home_goal: h, away_goal: a}), do: is_integer(h) and is_integer(a)

  @doc "Result from the home team's perspective: :home_win | :away_win | :draw | :unknown."
  def result(%__MODULE__{home_goal: h, away_goal: a}) when is_integer(h) and is_integer(a) do
    cond do
      h > a -> :home_win
      a > h -> :away_win
      true -> :draw
    end
  end

  def result(_), do: :unknown
end
